From 38a4fe1e8b3046d5170c6215307a6cff75c15f3f Mon Sep 17 00:00:00 2001
From: Benjamin Berg <bberg@redhat.com>
Date: Mon, 25 Jan 2021 11:26:58 +0100
Subject: [PATCH] uru4000: Call irq stop handler immediately if the transfer is
 cancelled

The irq handler may already be stopped if stop_irq_handler is called. In
that case, we should immediately call the handler rather than just never
calling it.

This fixes deactivation when the device is unexpectedly unplugged.

Closes: #355
---
 libfprint/drivers/uru4000.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/libfprint/drivers/uru4000.c b/libfprint/drivers/uru4000.c
index 0e67b2dc..f5d5d85b 100644
--- libfprint/drivers/uru4000.c
+++ libfprint/drivers/uru4000.c
@@ -375,6 +375,10 @@ stop_irq_handler (FpImageDevice *dev, irqs_stopped_cb_fn cb)
       g_cancellable_cancel (self->irq_cancellable);
       self->irqs_stopped_cb = cb;
     }
+  else
+    {
+      cb (dev);
+    }
 }
 
 /***** STATE CHANGING *****/
-- 
GitLab

